exec/java-exec/src/main/java/org/apache/drill/exec/util/filereader/DirectBufInputStream.java

```diff
@@ -0,0 +1,166 @@
+      long totalByteSize, boolean enableHints) {
+    super(in);
+    Preconditions.checkArgument(startOffset >= 0);
+    Preconditions.checkArgument(totalByteSize >= 0);
+    this.streamId = id;
+    this.allocator = allocator;
+    this.startOffset = startOffset;
+    this.totalByteSize = totalByteSize;
+    this.enableHints = enableHints;
+  }
+
+  public void init() throws IOException, UnsupportedOperationException {
+    checkStreamSupportsByteBuffer();
+    if (enableHints) {
+      fadviseIfAvailable(getInputStream(), this.startOffset, this.totalByteSize);
+    }
+    getInputStream().seek(this.startOffset);
+    return;
+  }
+
+  public int read() throws IOException {
+    return getInputStream().read();
+  }
+
+  public synchronized int read(DrillBuf buf, int off, int len) throws IOException {
+    buf.clear();
+    ByteBuffer directBuffer = buf.nioBuffer(0, len);
+    int lengthLeftToRead = len;
+    while (lengthLeftToRead > 0) {
+      lengthLeftToRead -= CompatibilityUtil.getBuf(getInputStream(), directBuffer, lengthLeftToRead);
+    }
+    buf.writerIndex(len);
+    return len;
+  }
+
+  public synchronized DrillBuf getNext(int bytes) throws IOException {
+    DrillBuf b = allocator.buffer(bytes);
+    int bytesRead = read(b, 0, bytes);
+    if (bytesRead <= -1) {
+      b.release();
+      return null;
+    }
+    return b;
+  }
+
+  public long getPos() throws IOException {
+    return getInputStream().getPos();
+  }
+
+  public boolean hasRemainder() throws IOException {
+    return getInputStream().available() > 0;
+  }
+
+  protected FSDataInputStream getInputStream() throws IOException {
+    // Make sure stream is open
+    checkInputStreamState();
+    return (FSDataInputStream) in;
+  }
+
+  protected void checkInputStreamState() throws IOException {
+    if (in == null) {
+      throw new IOException("Input stream is closed.");
+    }
+  }
+
+  protected void checkStreamSupportsByteBuffer() throws UnsupportedOperationException {
+    // Check input stream supports ByteBuffer
+    if (!(in instanceof ByteBufferReadable)) {
+      throw new UnsupportedOperationException("The input stream is not ByteBuffer readable.");
+    }
+  }
+
+  protected static void fadviseIfAvailable(FSDataInputStream inputStream, long off, long n) {
+    Method readAhead;
+    final Class adviceType;
+
+    try {
+      adviceType = Class.forName("org.apache.hadoop.fs.FSDataInputStream$FadviseType");
+    } catch (ClassNotFoundException e) {
+      logger.info("Unable to call fadvise due to: {}", e.toString());
+      readAhead = null;
+      return;
+    }
+    try {
+      Class<? extends FSDataInputStream> inputStreamClass = inputStream.getClass();
+      readAhead =
+          inputStreamClass.getMethod("adviseFile", new Class[] {adviceType, long.class, long.class});
+    } catch (NoSuchMethodException e) {
+      logger.info("Unable to call fadvise due to: {}", e.toString());
+      readAhead = null;
+      return;
+    }
```

Comment:
The above is relatively expensive. Should the work be done only once, then the method handle cached? If we are making many fadvise calls, then caching the method handle might provide a nice performance bump.